paper/paper.tex

```diff
@@ -751,13 +761,23 @@ \subsubsection{Julia is fast}
 
 As demonstrated in Section~\ref{sec:performance-comparison},
 Julia is not generically slower than traditional high-performance programming
-languages like C, C++, and Fortran. A minor exception is due to
-Julia being built on top of LLVM, which does not perform the same optimizations as other
-C/C++/Fortran compilers. In particular, LLVM does
-not use fused multiply-add (FMA) instructions by default, while GCC and Intel
-compilers do. Thus, one must opt-in to these optimizations explicitly.
-Relying on Julia's code manipulation techniques, this can be achieved with the macro
-\lstinline{@muladd}\footnote{\url{https://github.com/SciML/MuladdMacro.jl}}.
+languages like C, C++, and Fortran. A minor exception in the context of scientific
+computing relying on floating point operations is that Julia does not perform
+automatic fused multiply-add (FMA) contraction, that is replacing a multiplication
+followed by an addition with a single FMA instruction, as it is inconsistent
+with its strict floating point semantics. In other compilers, this optimization
+is controlled by a global compiler option (\lstinline{-ffp-contract} in GCC and
+Clang) and often enabled at reasonable optimization levels such as \lstinline{-O2}
+for GCC and Intel compilers\footnote{An interesting coincidence is that not only
+Julia but most LLVM-based frontends tend not to enable FMA contraction by default.}.
+In Julia, FMA contraction is controlled more locally: the
+\lstinline{muladd}\footnote{\url{https://docs.julialang.org/en/v1/base/math/\#Base.muladd}}
+intrinsic function allows the compiler to evaluate a multiply-add using the most
+efficient method. Relying on Julia's code manipulation techniques, the
+\lstinline{@muladd}\footnote{\url{https://github.com/SciML/MuladdMacro.jl}}
+macro makes it convenient to use the \lstinline{muladd} function by
+syntactically rewriting expressions to insert \lstinline{muladd} in appropriate
+locations.
```

Comment:
I like the new text and I find no fault with its content. However, I am wondering if we somewhat distort the message that Julia is fast, giving too much weight to the FMA discussion. If you look at the PDF, out of 17 lines in 4.1.1, 14 deal exclusively with FMA, compiler options and `muladd`:
![image](https://user-images.githubusercontent.com/3637659/148932321-aba3d239-37ca-4be1-84e4-238bbfe2ac34.png)

It might not be a major issue, but to a non-expert reader this might give the impression that this is a major issue for which complicated workarounds are required. What about only leaving an introductory statement and referring the inclined reader to a short appendix? That is, we could end this section with something like
> A minor exception in the context of scientific computing relying on floating point operations is that Julia does not perform automatic fused multiply-add (FMA) contraction, which can be remedied by explicitly using the `muladd` function or the more convenient `@muladd` macro (for more details see Appendix A).

If this kills the page limit, I'd also be ok to just end this subsection after "... convenient `@muladd` macro" and omitting the detailed discussion (although I find it highly instructive). What do you think?

Reply:
Good point :+1: I did that

Reply:
Great, thanks a lot!